src/ht.c

```diff
@@ -19,33 +30,85 @@ static const ut32 ht_primes_sizes[] = {
 	4166287, 4999559, 5999471, 7199369
 };
 
-static inline ut32 hashfn(SdbHt *ht, const void *k) {
-	return ht->hashfn ? ht->hashfn (k) : (ut32)(size_t)(k);
+static inline ut32 hashfn(SdbHt *ht, ut64 k) {
+	switch (ht->opt.type) {
+	case HT_TYPE_P:
+		return ht->opt.sub.p.hashfn ? ht->opt.sub.p.hashfn (ht_u2ptr (k)) : (ut32)(size_t)(k);
+	case HT_TYPE_U:
+		return ht->opt.sub.u.hashfn ? ht->opt.sub.u.hashfn (k) : (ut32)(size_t)(k);
+	case HT_TYPE_NONE:
+	default:
+		return ht->opt.sub.n.hashfn ? ht->opt.sub.n.hashfn (k) : (ut32)(size_t)(k);
+	}
 }
 
-static inline ut32 bucketfn(SdbHt *ht, const void *k) {
+static inline ut32 bucketfn(SdbHt *ht, ut64 k) {
 	return hashfn (ht, k) % ht->size;
 }
 
-static inline char *dupkey(SdbHt *ht, const void *k) {
-	return ht->dupkey ? ht->dupkey (k) : (char *)k;
+static inline ut64 dupkey(SdbHt *ht, ut64 k) {
+	switch (ht->opt.type) {
```

Comment:
i don't like this runtime type checking. thats not how generics work and i dont want to depend on -O3 to use a hashtable